rust/main/chains/hyperlane-sovereign/src/provider/rest_client.rs

```diff
@@ -1,179 +1,1127 @@
+            recipient: Some(H256::default()),
+            receipt: Some(TxnReceiptInfo {
+                gas_used: U256::default(),
+                cumulative_gas_used: U256::default(),
+                effective_gas_price: Some(U256::default()),
+            }),
+        };
+        Ok(res)
     }
 
-    pub async fn is_contract(&self) -> ChainResult<bool> {
-        todo!()
+    pub async fn get_batch(&self, batch: u64) -> ChainResult<Batch> {
+        info!("get_batch_tx_event(&self, batch: u64) batch:{:?}", batch);
+        let query = format!("/ledger/batches/{batch}?children=1");
+
+        let response = self
+            .http_get(&query)
+            .await
+            .map_err(|e| ChainCommunicationError::CustomError(format!("HTTP Get Error: {e}")))?;
+        let response: Schema<Batch> = serde_json::from_slice(&response)?;
+        let data = response.data.ok_or(ChainCommunicationError::CustomError(
+            "Invalid response".to_string(),
+        ))?;
+        Ok(data)
     }
 
-    pub async fn get_balance(&self) -> ChainResult<U256> {
-        todo!()
+    pub async fn get_tx_by_hash(&self, tx_id: String) -> ChainResult<Tx> {
+        info!("get_tx_by_hash(&self, tx_id: String) tx_id:{:?}", tx_id);
+        let query = format!("/ledger/txs/{tx_id}?children=1");
+
+        let response = self
+            .http_get(&query)
+            .await
+            .map_err(|e| ChainCommunicationError::CustomError(format!("HTTP Get Error: {e}")))?;
+        let response: Schema<Tx> = serde_json::from_slice(&response)?;
+        let data = response.data.ok_or(ChainCommunicationError::CustomError(
+            "Invalid response".to_string(),
+        ))?;
+        Ok(data)
     }
 
-    pub async fn get_chain_metrics(&self) -> ChainResult<Option<ChainInfo>> {
-        todo!()
+    // Return the latest slot, and the highest committed batch number in that slot.
+    pub async fn get_latest_slot(&self) -> ChainResult<(u32, Option<u32>)> {
+        info!("get_latest_slot(&self)");
+        #[derive(Clone, Debug, Deserialize)]
+        struct BatchRange {
+            // start: u32,
+            end: u32,
+        }
+        #[derive(Clone, Debug, Deserialize)]
+        struct Data {
+            batch_range: BatchRange,
+            number: u32,
+        }
+
+        let query = "/ledger/slots/latest?children=0";
+        let response = self
+            .http_get(query)
+            .await
+            .map_err(|e| ChainCommunicationError::CustomError(format!("HTTP Get Error: {e}")))?;
+        let response: Schema<Data> = serde_json::from_slice(&response)?;
+        let data = response.data.ok_or(ChainCommunicationError::CustomError(
+            "Invalid response".to_string(),
+        ))?;
+
+        // bach_range.end is exclusive - it's one above the last committed batch number
+        let last_batch = if data.batch_range.end > 0 {
+            Some(data.batch_range.end - 1)
+        } else {
+            None
+        };
```

Comment:
Would it be the same?

```suggestion
        let last_batch = data.batch_range.end.checked_sub(1);
```